implicits/src/com/rallyhealth/weepickle/v1/implicits/internal/Macros.scala

```diff
@@ -390,9 +390,17 @@ object Macros {
         """
 
         /**
-          * @see [[shouldDropDefault()]]
+          * Transforming the argument to itself applies any logic that the client may have provided
+          * in the implicit definition of From and/or To for the data type. This is important as this
+          * logic may transform the value into a default value which needs to be dropped.
+          *
+          * @see [[Argument.omitDefault]].
           */
-        if (arg.writingCheckDefault) q"""if (v.${TermName(arg.raw)} != ${arg.default}) $snippet"""
+        if (arg.writingCheckDefault)
+          q"""
+             val fromToArg = implicitly[${c.prefix}.FromTo[${arg.argType}]]
+             if (fromToArg.transform(v.${TermName(arg.raw)}, fromToArg) != ${arg.default}) $snippet
```

Comment:
There are cases where deep cloning objects will be expensive, e.g. `Some(deepTree)`. How can we avoid?

Reply:
I suppose we could have two different @dropDefault annotations: use our current annotation for having the default check based on the untransformed argument (current behavior) and a new one (say "@dropDefaultTransformed") that does the transform before checking (less surprising, but possibly more expensive behavior). 

However, I think the cases where the transform itself would be expensive would be quite rare because one would typically only use @dropDefault on fairly primitive types like `Option[String]`, so transforming would amount to little more than an object creation. So it may not be worth introducing that complexity.

Reply:
Yeah, `Option[String]` would be relatively cheap. I'm thinking of:
```scala
case class Foo(@dropDefault b: List[Bar] = Nil)
case class Bar(@dropDefault q: List[Baz] = Nil)
case class Quux(@dropDefault e: List[Expensive] = Nil)
case class Expensive(large: JsObject)
```
After this PR, I'd expect `FromScala(foo).transform(to)` would make `len(b) * len(q) * len(e)` copies of the `JsObject`.  (Multiply again by two if we only did a naive implementation of `length`.) Making the expensive behavior the default one seems surprising as well.

Let's step back for a sec and look at the use case. From the tests, it looks like we want `Foo(bar = Some(""))` to become `{}`. How did we get an input of `Some("")`? Is there a reason that the transformation of `Some("") => None` cannot be done in the source?

> And I'm totally open to other suggestions.

One solution might be annotating secondary values that should be normalized to/from the default, e.g. `@alsoDefaultValues(Some(""))`. Comparing static values is always cheap.

If we need dynamic mappings, I guess maybe we maybe we could add a `Normalizer[T]` type class, but it'd be weird that it works for values in a case class, but not a `Map`. I'm uneasy about that direction.

Another approach for dynamic values would be to have a Visitor strip `field: null` out of the payloads. MessagePack [needs to know map length upfront](https://github.com/msgpack/msgpack/blob/master/spec.md#map-format-family), so it'd probably belong somewhere near `ToJson`. Maybe `ToNullStrippedJson`? `ToJson.map(visitor => new DeepNullFieldRemovingVisitor(vistior))`?

Reply:
Annotating secondary values sounds good on the surface, but you are replicating logic that should be applied implicitly. What if that logic itself is expensive -- then you will be doing that logic twice. Feels like current `dropDefault` is really `dropFromDefault`, where what I am recommending is `dropToDefault` (just makes more intuitive sense to me). But I'd be okay adding a new annotation for `dropToDefault` and then letting the clients decide what behavior they want (recognizing `dropToDefault` is a terrible name). That way one could have something like:
```scala
implicit val myWeirdBarCalculator FromTo[List[Bar]] = ???
 
case class Foo(
  @dropToDefault b: List[Bar] = Nil
  @dropDefault e: List[Expensive] = Nil
)
```

Reply:
I'm going to try an implementation by adding `@dropDefaultWithTranslation` (thinking that's a better name) and address the inconsistency on line 413, and see what that looks like.

Reply:
Admittedly, yours is more DRY.

Do you think people will understand when to use each and the associated costs? I sometimes forget to `@dropDefault`, myself. Exponential complexity over depth is kinda scary.

I'd love a solution that never has exponential cost. For example, the only thing we really need to apply is the `MapFrom.f: U => T`. The rest is waste.